### DISCUSSION TITLE
Programmable Circle Crop on Gif

### DISCUSSION THREAD

**Discussion by tapitchforth** [ImageMagick/ImageMagick#2702] [Help] [answered]
Hello,

I am writing a batch script to perform a bunch of processing using IM and Gifsicle however trying to perform a circular crop of an image has me stumped.  I've looked through a bunch of different articles and have some working code, but I can't seem to make the circle size programmable and keep it in the middle of the Gif for the crop - it keeps defaulting to the top left corner unless I make the circle fill the image.

The code I have is:
```
set diameter=180
set input=input_256x256.gif
set output=circular_cropped_256x256.gif
set /a radius=%diameter% / 2

magick convert ^
  %input% -layers coalesce -alpha on NULL: ^
  ( -clone 0 -channel a -evaluate multiply 0 +channel -fill white -draw "circle 64 64 64 1" -repage +0+0 ) ^
  -compose DstIn -layers composite %output%
```

Ideally I would like this to support any input size, diameter (user programmable) and offset (user programmable) however just getting the circle centred on a 256x256 gif is my priority at the moment.

Any advice on why the code I have isn't working (and ideally some solution tips) would be greatly appreciated 👍

**Comment by snibgo** [comment, marked as answer]
I suppose your IM is v7. I suggest you use "magick", not "magick convert".

Your difficulty seems to be in the specification of the circle. This needs two coordinate-pairs. The first is of the circle centre, and the second is any pixel on the edge.

So, if the image is 256x256 pixels, so might have:
```
-draw "circle 128,128 10,128"
```
This circle is centred in the image, and nearly reaches the edges of the image.

Instead of numbers, you can use %[fx:...] expressions. For example:
```
set OFFSX=10
set OFFSY=20
set RAD=50
:
:

-draw "circle %[fx:w/2+%OFFSX%],%[fx:h/2+%OFFSY%] %[fx:w/2+%OFFSX%],%[fx:h/2+%OFFSY%-%RAD%]" x.png
```

**Reply by tapitchforth** [reply]
Thanks for the detailed reply, the new co-ordinates & formulas you provided sorted this!

Greatly appreciated!

**Reply by snibgo** [reply]
As GeeMack suggests below, it is more accurate to subtract 1 from w and h, the width and height, like this:
```
-draw "circle %[fx:(w-1)/2+%OFFSX%],%[fx:(h-1)/2+%OFFSY%] %[fx:(w-1)/2+%OFFSX%],%[fx:(h-1)/2+%OFFSY%-%RAD%]"
```

**Comment by GeeMack** [comment]
With ImageMagick v7 you don't normally want to use "convert" unless you need to access particular behavior from IMv6. Use just "magick".

Try including some FX expressions in your "-draw" operation to make it work with variable sized images...

     ... -draw "circle %[fx:(w-1)/2],%[fx:(h-1)/2] %[fx:(w-1)/2],1" ...

That expression "(w-1)/2" is actually the exact center of the width for your "-draw" operation.

If you're running it from a BAT script make sure to double the percent signs "%%" on those FX expressions.

**Reply by tapitchforth** [reply]
I've removed the "convert" part of the command and also popped in some double % based on your example with FX expression.

Many thanks for all of the advice, super helpful!
